src/js/protocols/webusbdfu.js

```diff
@@ -0,0 +1,1119 @@
+        .catch((e) => {
+            console.error(`There is no device. ${e}`);
+        });
+    }
+    openDevice() {
+        this.usbDevice
+        .open()
+        .then(async () => {
+            // show key values for the device
+            console.log(`USB Device opened: ${this.usbDevice.productName}`);
+            if (this.usbDevice.configuration === null) await this.usbDevice.selectConfiguration(1);
+            this.claimInterface(0);
+        })
+        .catch(error => {
+            console.log('Failed to open USB device:', error);
+            gui_log(i18n.getMessage('usbDeviceOpenFail'));
+        });
+    }
+    closeDevice() {
+        this.usbDevice
+        .close()
+        .then(() => {
+            gui_log(i18n.getMessage('usbDeviceClosed'));
+            console.log('DFU Device closed');
+        })
+        .catch(error => {
+            console.log('Failed to close USB device:', error);
+            gui_log(i18n.getMessage('usbDeviceCloseFail'));
+        });
+        this.usbDevice = null;
+    }
+    claimInterface(interfaceNumber) {
+        const self = this;
+
+        self.usbDevice
+        .claimInterface(interfaceNumber)
+        .then(() => {
+            console.log(`Claimed interface: ${interfaceNumber}`);
+
+            if (self.options.exitDfu) {
+                self.leave();
+            } else {
+                self.upload_procedure(0);
+            }
+        })
+        .catch(error => {
+            console.log('Failed to claim USB device', error);
+            self.cleanup();
+        });
+    }
+    releaseInterface(interfaceNumber) {
+        this.usbDevice
+        .releaseInterface(interfaceNumber, function released() {
+            console.log(`Released interface: ${interfaceNumber}`);
+            this.closeDevice();
+        })
+        .catch(error => {
+            console.log(`Could not release interface: ${interfaceNumber} (${error})`);
+        });
+    }
+    resetDevice(callback) {
+        this.usbDevice
+        .reset()
+        .then(() => {
+            console.log('Reset Device');
+            callback?.();
+        })
+        .catch(error => {
+            console.log(`Could not reset device: ${error}`);
+            callback?.();
+        });
+    }
+    getString(index, callback) {
+        const setup = {
+            requestType: 'standard',
+            recipient: 'device',
+            request: 6,
+            value: 0x300 | index,
+            index: 0, // specifies language
+        };
+
+        this.usbDevice
+        .controlTransferIn(setup, 255)
+        .then(result => {
+            if (result.status === 'ok') {
+                const _length = result.data.getUint8(0);
+                let _descriptor = '';
+                for (let i = 2; i < _length; i += 2) {
+                    const charCode = result.data.getUint16(i, true);
+                    _descriptor += String.fromCharCode(charCode);
+                }
+                callback(_descriptor, 0);
```

Comment:
I'm not familiar with usb so have a noob question probably.

What populates `result.data`? Like do we write into that thing, are first 8 bits which represent the length something we write or something standard?

Asking because we could potentially avoid the whole `for` loop here for string parsing and use [TextDecoder](https://developer.mozilla.org/en-US/docs/Web/API/TextDecoder) instead.

Even if we just cut off the `length` and pass in the rest text decoder should be able to handle that.

Reply:
`controlTransferIn` returns a DataView object containing the data received from the USB device, if any.

![image](https://github.com/betaflight/betaflight-configurator/assets/8344830/5b9db02d-3820-40d8-b511-e90eece0ab44)